### PR TITLE
Add mailcow OAuth

### PR DESCRIPTION
**Pull request by thisni1s**
I based this on the Patroen PR #3323 and the  Yandex commit 435eca6 .
I will try to add and or fix anything that is missing.
I also tested this with the pocketbase JS SDK and have run `make test` and `make lint`

### CODE REVIEW

**Comment by ganigeorgiev** [comment]
Thank you for the PR but I'm not sure about this integration.

As far as I understand this is a self-hosted only service, right?

Keep in mind that I'm not familiar with mailcow and their documentation lack details about how the OAuth2 is implemented and what flows are supported (eg. PKCE).

I'm also little confused whether this is actually needed because it seems that based on the following PR https://github.com/mailcow/mailcow-dockerized/pull/5381 and the blog post in https://mailcow.email/posts/2023/ldap-announcement/, the mailcow team seems to plan adding support for OIDC. Here is an excerpt from the LDAP blogpost:

> ... **The mailcow is connected to Keycloak via OIDC** ...

_(if I'm reading it correctly they'll add OIDC support to allow Keycloack connecting to mailcow and then use Keyclock for other auth integration like LDAP)_


But the above PR and this OIDC implementation doesn't seem to be merged yet (there is a PR to what they call "nightly" in https://github.com/mailcow/mailcow-dockerized/pull/5387 but I haven't tested it).

---

So at the end, if they ended up adding support for OIDC, the current custom implementation no longer will be needed as PocketBase devs can use the existing generic OIDC provider.

I'll close the PR for now to avoid the cluttering, but feel to correct me if I'm reading the above resources incorrectly and a custom mailcow integration would be definitely required.

**Comment by thisni1s** [comment]
Hi,
yes mailcow is a self-hosted service only and i made this integration to serve my specific needs :)

Their OAuth2 integration really is badly documented, but it is based upon this PHP library https://bshaffer.github.io/oauth2-server-php-docs/

They will add OIDC support for mailcow, where mailcow acts as the OIDC Client and not as a provider.
So without the direct integration you would have to set up an identity management service like keycloak or authentik as the source of the identities and then have mailcow and pocketbase be the clients of this service instead of having mailcow as the provider and managing all the accounts there.
So their OIDC support would not enable pocketbase users to connect to mailcow without using an additional identity management software, thats why i felt a direct integration would be easier.

Still in theory it will be possible with their OIDC implementation to connect everything together, but you have to use an additional service.

Maybe i could clear some things up, in the end its up to you to decide if the direct integration is worth it or not.

**Review by ganigeorgiev — CHANGES_REQUESTED** [review]
In this case I guess then it makes sense to have a custom provider.

I've added some nitpicks but keep in mind that this PR remains with a low priority (_it will be merged eventually in the `develop` branch_) and I'll have to find time to setup a local mailcow instance to test it (_the client-side `OIDCOptions` and `SelfHostedOptions` could also be merged in a more generic component but I'll refactor that later_).

File: tools/auth/mailcow.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		Active      int    `json:"active"`
+	}{}
+	if err := json.Unmarshal(data, &extracted); err != nil {
+		return nil, err
+	}
+
+	// Mailcow usernames are usually just the email adresses, so we just take the part in front of the @
+	if strings.Contains(extracted.Username, "@") {
+		extracted.Username = strings.Split(extracted.Username, "@")[0]
+	}
```
[the remark this answers is not in the document]
Comment (thisni1s) [reply tools/auth/mailcow.go]: If the mailcow user has a mailbox, which is the case 99% of the time, the username/id/identifier fields are the email address even though the array has two different keys

**Comment by ganigeorgiev** [comment]
@thisni1s Could you provide some guidance/steps how to create and test the mailcow OAuth2 app?

**Comment by ganigeorgiev** [comment]
I've merged the PR with some minor nitpicks in the `develop` branch.

I figured out how to test it  - the OAuth2 app settings were buried under _System > Configuration > Access tab/dropdown > OAuth2 Apps_.

For the test I've used the live demo https://nightly-demo.mailcow.email/. There are some issues with its certificate as it is not recognized for some reason as part of the root chain, but other than that the integration seems to work fine.

The provider will be part of the next PocketBase v0.19.0 release soon.

**Comment by thisni1s** [comment]
Perfect, thank you

**Pull request closed without merging**